viz/metrics-api/client/client.go

```diff
@@ -0,0 +1,200 @@
+	var msg pb.ListPodsResponse
+	err := c.apiRequest(ctx, "ListPods", req, &msg)
+	return &msg, err
+}
+
+func (c *grpcOverHTTPClient) ListServices(ctx context.Context, req *pb.ListServicesRequest, _ ...grpc.CallOption) (*pb.ListServicesResponse, error) {
+	var msg pb.ListServicesResponse
+	err := c.apiRequest(ctx, "ListServices", req, &msg)
+	return &msg, err
+}
+
+func (c *grpcOverHTTPClient) Tap(ctx context.Context, req *pb.TapRequest, _ ...grpc.CallOption) (pb.Api_TapClient, error) {
+	return nil, status.Error(codes.Unimplemented, "Tap is deprecated in public API, use tap APIServer")
+}
+
+func (c *grpcOverHTTPClient) TapByResource(ctx context.Context, req *pb.TapByResourceRequest, _ ...grpc.CallOption) (pb.Api_TapByResourceClient, error) {
+	return nil, status.Error(codes.Unimplemented, "Tap is deprecated in public API, use tap APIServer")
+}
+
+func (c *grpcOverHTTPClient) apiRequest(ctx context.Context, endpoint string, req proto.Message, protoResponse proto.Message) error {
+	url := c.endpointNameToPublicAPIURL(endpoint)
+
+	log.Debugf("Making gRPC-over-HTTP call to [%s] [%+v]", url.String(), req)
+	httpRsp, err := c.post(ctx, url, req)
+	if err != nil {
+		return err
+	}
+	defer httpRsp.Body.Close()
+	log.Debugf("gRPC-over-HTTP call returned status [%s] and content length [%d]", httpRsp.Status, httpRsp.ContentLength)
+
+	if err := protohttp.CheckIfResponseHasError(httpRsp); err != nil {
+		return err
+	}
+
+	reader := bufio.NewReader(httpRsp.Body)
+	return protohttp.FromByteStreamToProtocolBuffers(reader, protoResponse)
+}
+
+func (c *grpcOverHTTPClient) post(ctx context.Context, url *url.URL, req proto.Message) (*http.Response, error) {
+	reqBytes, err := proto.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	httpReq, err := http.NewRequest(
+		http.MethodPost,
+		url.String(),
+		bytes.NewReader(reqBytes),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	rsp, err := c.httpClient.Do(httpReq.WithContext(ctx))
+	if err != nil {
+		log.Debugf("Error invoking [%s]: %v", url.String(), err)
+	} else {
+		log.Debugf("Response from [%s] had headers: %v", url.String(), rsp.Header)
+	}
+
+	return rsp, err
+}
+
+func (c *grpcOverHTTPClient) endpointNameToPublicAPIURL(endpoint string) *url.URL {
+	return c.serverURL.ResolveReference(&url.URL{Path: endpoint})
+}
+
+func newClient(apiURL *url.URL, httpClientToUse *http.Client, namespace string) (pb.ApiClient, error) {
+	if !apiURL.IsAbs() {
+		return nil, fmt.Errorf("server URL must be absolute, was [%s]", apiURL.String())
+	}
+
+	serverURL := apiURL.ResolveReference(&url.URL{Path: APIPrefix})
+
+	log.Debugf("Expecting API to be served over [%s]", serverURL)
+
+	return &grpcOverHTTPClient{
+		serverURL:  serverURL,
+		httpClient: httpClientToUse,
+		namespace:  namespace,
+	}, nil
+}
+
+// NewInternalClient creates a new Viz API client intended to run inside a
+// Kubernetes cluster.
+func NewInternalClient(namespace string, kubeAPIHost string) (pb.ApiClient, error) {
```

Comment:
May be worth nothing this and `NewExternalClient` don't have that dance I commented above in public API with one function wrapping another but not for any particular reason.